extensions/wikia/WikiFactory/SpecialWikiFactory_body.php

```diff
@@ -61,160 +65,201 @@ public function execute( $subpage ) {
 		/**
 		 * initial output
 		 */
-		$wgOut->setPageTitle( wfMsg('wikifactorypagetitle') );
-		$wgOut->setRobotpolicy( 'noindex,nofollow' );
+		$wgOut->setPageTitle( wfMsg( 'wikifactorypagetitle' ) );
+		$wgOut->setRobotPolicy( 'noindex,nofollow' );
 		$wgOut->setArticleRelated( false );
 
-		if ( in_array( strtolower($subpage), array( "metrics", "metrics/main", "metrics/monthly", "metrics/daily" ) ) ) {
+		if ( in_array( strtolower( $subpage ), self::METRIC_PATHS ) ) {
 			$oAWCMetrics = new WikiMetrics();
 			$oAWCMetrics->show( $subpage );
-		}
-		elseif ( strpos( $subpage, "short.stats" ) === 0 ) {
+		} elseif ( strpos( $subpage, "short.stats" ) === 0 ) {
 			$subpageOptions = explode( '/', $subpage );
 			$lang = isset( $subpageOptions[1] ) ? $subpageOptions[1] : null;
 			$wgOut->addHTML( $this->shortStats( $lang ) );
-		}
-		elseif ( strpos( $subpage, "long.stats" ) === 0 ) {
+		} elseif ( strpos( $subpage, "long.stats" ) === 0 ) {
 			$subpageOptions = explode( '/', $subpage );
 			$lang = isset( $subpageOptions[1] ) ? $subpageOptions[1] : null;
 			$wgOut->addHTML( $this->longStats( $lang ) );
-		}
-		elseif ( strtolower($subpage) === "add.variable" ) {
-			$varOverrides = array();
-			$wgOut->addHTML( $this->doAddVariableForm($varOverrides) ); // process the post (if relevant).
-			$wgOut->addHTML( $this->addVariableForm($varOverrides) ); // display the form
-		}
-		else {
+		} elseif ( strtolower( $subpage ) === "add.variable" ) {
+			$varOverrides = [];
+			$wgOut->addHTML( $this->doAddVariableForm( $varOverrides ) ); // process the post (if relevant).
+			$wgOut->addHTML( $this->addVariableForm( $varOverrides ) ); // display the form
+		} else {
 			$subpage = ( $subpage == "/" ) ? null : $subpage;
-			$oWiki = $this->getWikiData( $subpage );
+			$this->handleNavigationTab( $subpage );
+			$cityId = $this->findCityId( $subpage );
 
-			if( !isset( $oWiki->city_id )) {
+			if ( empty( $cityId ) ) {
 				$this->doWikiSelector();
+				return;
 			}
-			else {
-				$this->mWiki = $oWiki;
-				$this->mTags = new WikiFactoryTags( $this->mWiki->city_id );
-				$this->doWikiForm( );
+
+			$this->mTitle = Title::makeTitle(
+				NS_SPECIAL,
+				implode( '/', [ 'WikiFactory', $cityId, $this->mTab ] )
+			);
+
+			if ( !preg_match("/$cityId/", $subpage ) ) {
+				$wgOut->redirect( $this->mTitle->getLocalURL() );
+				return;
 			}
+
+			$wiki = $this->getWikiData( $subpage );
+
+			$this->mWiki = $wiki;
+			$this->mTags = new WikiFactoryTags( $this->mWiki->city_id );
+
+			$this->doWikiForm();
 		}
 	}
 
 	/**
 	 * getWikiData
 	 *
-	 * use subpage as param and try to find wiki which match criteria
-	 *
-	 * @access private
+	 * Use subpage as param and try to find wiki which match criteria
 	 *
 	 * @param mixed $subpage:
 	 *
 	 * @return mixed: database row from city_list
 	 */
 	private function getWikiData( $subpage ) {
```

Comment:
This method did waaaaaay more than just get wiki data.  Updated and split out the extraneous functionality.